### PR TITLE
Donation improvements

### LINKED ISSUE #220: /admin/donation should allow editing of fields
mostly date paid, ship date (also for pickup date, eventually), and comments.

### LINKED ISSUE #278: Premiums should be disabled when not in radiothon mode!
this prevents non-radiothon donors from filling out extraneous fields and from thinking they'll get premiums.

### PR DESCRIPTION
**Pull request by echarlie**
fixes #278
partially resolves #220

### CODE REVIEW

File: wuvt/templates/admin/donation_index.html [hunk 1]
```diff
@@ -62,7 +67,12 @@ <h1>Donations</h1>
                 {{ donation.city }}, {{ donation.state }} {{ donation.zipcode }}
 {% endif %}
             </td>
-            <td>{% if donation.shipped_date %}{{ donation.shipped_date|datetime }} {% endif %}</td>
+            <td>{% if donation.shipped_date %}{{ donation.shipped_date|datetime }}
+		{% else %}<form method="post">
+			<input type="hidden" name="_csrf_token" value="{{ csrf_token() }}">
+			<input type="hidden" name="id" value="{{ donation.id }}">
+			<button name="set_shipped" type="submit" class="btn btn-sm">Set as Shipped</button></form>
+		{% endif %}{% endif %}</td>
```
Comment (mutantmonkey) [inline wuvt/templates/admin/donation_index.html:75]: I'm assuming this is a typo. This is causing the following error:
```
jinja2.exceptions.TemplateSyntaxError: Encountered unknown tag 'endif'. Jinja was looking for the following tags: 'endfor' or 'else'. The innermost block that needs to be closed is 'for'.
```

File: wuvt/admin/views.py [hunk 2]
```diff
@@ -492,6 +492,29 @@ def donation_index():
...
+            try:
+                db.session.commit()
+            except:
+                db.session.rollback()
+                raise
+
+        # TODO support setting asynchronously, rather than redirecting
+        if 'set_shipped' in request.form:
+            # TODO log who set as paid
+            o = Order.get_or_404(id=request.form["id"])
```
Comment (mutantmonkey) [inline wuvt/admin/views.py:509]:
```
AttributeError: type object 'Order' has no attribute 'get_or_404'
```

This should be `o = Order.query.get(request.form["id"])`

File: wuvt/admin/views.py [hunk 3]
```diff
@@ -492,6 +492,29 @@ def donation_index():
         if 'reset_stats' in request.form:
             redis_conn.set('donation_stats_start',
                            datetime.datetime.utcnow().isoformat())
+        if 'set_paid' in request.form:
+            # TODO log who set as paid
+            o = Order.get_or_404(id=request.form["id"])
```
Comment (mutantmonkey) [inline wuvt/admin/views.py:497]:
```
AttributeError: type object 'Order' has no attribute 'get_or_404'
```

This should be `o = Order.query.get(request.form["id"])`

File: wuvt/templates/admin/donation_index.html (hunk 1 above)
Comment (mutantmonkey) [reply wuvt/templates/admin/donation_index.html:75]: I also made this change and the button seems to still be showing after I click. Not sure if it's not updating the database or not but it doesn't work.

**Review by mutantmonkey — CHANGES_REQUESTED** [review]

File: wuvt/admin/views.py (hunk 2 above)
Comment (echarlie) [reply wuvt/admin/views.py:509]: ah, right. or `.query.get_or_404`, but that may not make sense here...

File: wuvt/admin/views.py [hunk 4]
```diff
@@ -492,6 +492,29 @@ def donation_index():
...
+            except:
+                db.session.rollback()
+                raise
+
+        # TODO support setting asynchronously, rather than redirecting
+        if 'set_shipped' in request.form:
+            # TODO log who set as paid
+            o = Order.query.get(request.form["id"])
+            o.set_shipped()
+            db.session.add(o)
```
Comment (mutantmonkey) [inline wuvt/admin/views.py:511]: What is being added here? This call doesn't seem correct; removing it should work.

File: wuvt/admin/views.py [hunk 5]
```diff
@@ -492,6 +492,29 @@ def donation_index():
         if 'reset_stats' in request.form:
             redis_conn.set('donation_stats_start',
                            datetime.datetime.utcnow().isoformat())
+        if 'set_paid' in request.form:
+            # TODO log who set as paid
+            o = Order.query.get(request.form["id"])
+            o.set_paid(method="check")
+            db.session.add(o)
```
Comment (mutantmonkey) [inline wuvt/admin/views.py:499]: What is being added here? This call doesn't seem correct; removing it should work.

**Review by mutantmonkey — CHANGES_REQUESTED** [review]
It seems like marking an order as shipped doesn't work if there was no shipping address entered. I'm not sure if this is because `db.session.add` is being called erroneously or if it's something else. It does seem to work if there is a shipping address, though.

File: wuvt/admin/views.py (hunk 4 above)
Comment (echarlie) [reply wuvt/admin/views.py:511]: is my memory of sqlalchemy way off? I thought any db modification required you add it to the session? Or is it implicity added when you read from the DB?

Comment (echarlie) [reply wuvt/admin/views.py:511]: nevermind. I read the docs. indeed.

**Comment by echarlie** [comment]
interesting quirk. We don't have any DB constraints that would cause this problem, do we?

**Comment by mutantmonkey** [comment]
Not that I can think of...

**Comment by mutantmonkey** [comment]
Okay, I think this is why:
```python
    def set_shipped(self, shipped_date=None):
        if self.premiums is not None:
```

If premiums is `None`, `set_shipped` does nothing; in this case, we should hide the "Set as Shipped" button.

Looks good otherwise, this will be a great improvement.

File: wuvt/admin/views.py [hunk 6]
```diff
@@ -492,6 +492,27 @@ def donation_index():
         if 'reset_stats' in request.form:
             redis_conn.set('donation_stats_start',
                            datetime.datetime.utcnow().isoformat())
+        if 'set_paid' in request.form:
+            # TODO log who set as paid
```
Comment (mutantmonkey) [inline wuvt/admin/views.py:496]: I don't love leaving these TODOs in the source code with nothing else to keep track of them; they're easy to forget. Can you create an issue for this one please?

File: wuvt/admin/views.py [hunk 7]
```diff
@@ -492,6 +492,27 @@ def donation_index():
...
+            o = Order.query.get(request.form["id"])
+            o.set_paid(method="check")
+            try:
+                db.session.commit()
+            except:
+                db.session.rollback()
+                raise
+
+        # TODO support setting asynchronously, rather than redirecting
```
Comment (mutantmonkey) [inline wuvt/admin/views.py:505]: This one is less important, not sure we need an issue for this right now.

File: wuvt/donate/models.py [hunk 8]
```diff
@@ -70,3 +70,11 @@ def set_paid(self, method, paid_date=None):
             self.paid_date = paid_date
         else:
             self.paid_date = datetime.datetime.utcnow()
+
+    def set_shipped(self, shipped_date=None):
+        if self.premiums is not None:
```
Comment (mutantmonkey) [inline wuvt/donate/models.py:75]: So if premiums is `None`, then this function does nothing. Might be a good idea to instead return an error.

**Review by mutantmonkey — CHANGES_REQUESTED** [review]

**Comment by echarlie** [comment]
discussion should continue in wuvt/donormotor#22

**Pull request closed without merging**